File: plugins/plugin-prose/rust/src/actions/compile.rs
//! PROSE_COMPILE action for validating OpenProse programs

use async_trait::async_trait;
use regex::Regex;
use serde_json::{json, Value};
use std::path::Path;
use tracing::info;

use crate::generated::specs::require_action_spec;
use crate::services::ProseService;
use crate::{Action, ActionExample, ActionResult};

/// Extract a value from an XML tag
fn extract_xml_value(text: &str, tag: &str) -> Option<String> {
    let pattern = format!(r"(?i)<{}>([\s\S]*?)</{}>", tag, tag);
    let re = Regex::new(&pattern).ok()?;
    re.captures(text)
        .and_then(|caps| caps.get(1))
        .map(|m| m.as_str().trim().to_string())
}

/// Basic validation result
struct ValidationResult {
    valid: bool,
    errors: Vec<String>,
    warnings: Vec<String>,
    summary: String,
}

/// Perform basic validation on prose content
fn basic_validate(content: &str) -> ValidationResult {
    let mut errors = Vec::new();
    let mut warnings = Vec::new();
    let lower = content.to_lowercase();

    // Check for program block
    if !lower.contains("program") {
        errors.push("Missing program declaration".to_string());
    }

    // Check for balanced braces
    let open_braces = content.matches('{').count();
    let close_braces = content.matches('}').count();
    if open_braces != close_braces {
        errors.push(format!(
            "Unbalanced braces: {} open, {} close",
            open_braces, close_braces
        ));
    }

    // Check for session definition
    if !lower.contains("session") {
        warnings.push("No session defined - program may not have an entry point".to_string());
    }

    // Check for version
    if !lower.contains("version") {
        warnings.push("No version specified".to_string());
    }

    let valid = errors.is_empty();
    let summary = if valid {
        "Program is syntactically correct.".to_string()
    } else {
        "Program has validation errors.".to_string()
    };

    ValidationResult {
        valid,
        errors,
        warnings,
        summary,
    }
}

/// Action to validate an OpenProse program
pub struct ProseCompileAction {
    name: &'static str,
    description: &'static str,
    similes: Vec<&'static str>,
    examples: Vec<ActionExample>,
}

impl ProseCompileAction {
    pub fn new() -> Self {
        let spec = require_action_spec("PROSE_COMPILE");
        Self {
            name: spec.name,
            description: spec.description,
            similes: spec.similes.clone(),
            examples: spec
                .examples
                .iter()
                .map(|ex| ActionExample {
                    user_message: ex[0].1.to_string(),
                    agent_response: ex[1].1.to_string(),
                })
                .collect(),
        }
    }

    fn extract_file(&self, text: &str) -> Option<String> {
        if let Some(file) = extract_xml_value(text, "file") {
            return Some(file);
        }

        let lower = text.to_lowercase();

        // "prose compile <file>" or "prose validate <file>"
        let re = Regex::new(r"prose\s+(?:compile|validate)\s+(\S+)").ok()?;
        if let Some(caps) = re.captures(&lower) {
            return Some(caps.get(1)?.as_str().to_string());
        }

        // "check <file.prose>" or "validate <file.prose>"
        let re = Regex::new(r"(?:check|validate)\s+(\S+\.prose)").ok()?;
        if let Some(caps) = re.captures(&lower) {
            return Some(caps.get(1)?.as_str().to_string());
        }

        None
    }
}

impl Default for ProseCompileAction {
    fn default() -> Self {
        Self::new()
    }
}

#[async_trait]
impl Action for ProseCompileAction {
    fn name(&self) -> &str {
        self.name
    }

    fn similes(&self) -> Vec<&str> {
        self.similes.clone()
    }

    fn description(&self) -> &str {
        self.description
    }

    async fn validate(&self, message: &Value, _state: &Value) -> bool {
        let content = message
            .get("content")
            .and_then(|c| c.get("text"))
            .and_then(|t| t.as_str())
            .unwrap_or("");

        let lower = content.to_lowercase();

        lower.contains("prose compile")
            || lower.contains("prose validate")
            || (lower.contains("check") && lower.contains(".prose"))
            || (lower.contains("validate") && lower.contains(".prose"))
    }

    async fn handler(
        &self,
        message: &Value,
        _state: &Value,
        service: Option<&mut ProseService>,
    ) -> ActionResult {
        let content = message
            .get("content")
            .and_then(|c| c.get("text"))
            .and_then(|t| t.as_str())
            .unwrap_or("");

        let file = match self.extract_file(content) {
            Some(f) => f,
            None => {
                return ActionResult {
                    success: false,
                    text: "Please specify a .prose file to validate. Example: `prose compile workflow.prose`".to_string(),
                    data: None,
                    error: Some("No file specified".to_string()),
                };
            }
        };

        let svc = match service {
            Some(s) => s,
            None => {
                return ActionResult {
                    success: false,
                    text: "Prose service not available".to_string(),
                    data: None,
                    error: Some("Service unavailable".to_string()),
                };
            }
        };

        let cwd = std::env::current_dir().unwrap().display().to_string();
        let file_path = if Path::new(&file).is_absolute() {
            file.clone()
        } else {
            Path::new(&cwd).join(&file).display().to_string()
        };

        if !svc.file_exists(&file_path).await {
            return ActionResult {
                success: false,
                text: format!("File not found: {}", file_path),
                data: None,
                error: Some("File not found".to_string()),
            };
        }

        let program_content = match svc.read_prose_file(&file_path).await {
            Ok(c) => c,
            Err(e) => {
                return ActionResult {
                    success: false,
                    text: format!("Failed to read file: {}", e),
                    data: None,
                    error: Some(e.to_string()),
                };
            }
        };

        let result = basic_validate(&program_content);

        info!("Validated {}: valid={}", file, result.valid);

        let mut parts = Vec::new();
        parts.push(format!("## Validation Results for {}\n", file));
        parts.push(format!(
            "**Status:** {}\n",
            if result.valid { "✓ Valid" } else { "✗ Invalid" }
        ));
        parts.push(format!("**Summary:** {}\n", result.summary));

        if !result.errors.is_empty() {
            parts.push("\n### Errors\n".to_string());
            for error in &result.errors {
                parts.push(format!("- ❌ {}", error));
            }
        }

        if !result.warnings.is_empty() {
            parts.push("\n### Warnings\n".to_string());
            for warning in &result.warnings {
                parts.push(format!("- ⚠️ {}", warning));
            }
        }

        if result.valid && result.errors.is_empty() && result.warnings.is_empty() {
            parts.push("\nNo issues found. Program is ready to run.".to_string());
        }

        ActionResult {
            success: true,
            text: parts.join("\n"),
            data: Some(json!({
                "valid": result.valid,
                "errors": result.errors,
                "warnings": result.warnings,
                "file": file,
            })),
            error: None,
        }
    }

    fn examples(&self) -> Vec<ActionExample> {
        self.examples.clone()
    }
}

File: plugins/plugin-prose/typescript/actions/compile.ts
import * as path from "node:path";
import type { Action, HandlerCallback, IAgentRuntime, Memory, State } from "@elizaos/core";
import { generateText, logger, ModelType } from "@elizaos/core";
import { requireActionSpec } from "../generated/specs/specs";
import { createProseService, type ProseService } from "../services/proseService";
import type { ProseCompileResult } from "../types";

const spec = requireActionSpec("PROSE_COMPILE");

// Service cache
const serviceCache = new WeakMap<IAgentRuntime, ProseService>();

function getService(runtime: IAgentRuntime): ProseService {
  let service = serviceCache.get(runtime);
  if (!service) {
    service = createProseService(runtime, {});
    serviceCache.set(runtime, service);
  }
  return service;
}

const EXTRACT_TEMPLATE = `<extraction_task>
Extract the prose file path from the user's message.

User message: {{content}}

Respond with ONLY this XML tag:
<file>path/to/program.prose</file>

If no file path is found, leave the tag empty.
</extraction_task>`;

const COMPILE_TEMPLATE = `You are the OpenProse compiler/validator. Analyze the following .prose program
for syntax errors, structural issues, and potential problems.

{{compiler_spec}}

═══════════════════════════════════════════════════════════════
                    PROGRAM TO VALIDATE
═══════════════════════════════════════════════════════════════

\`\`\`prose
{{program}}
\`\`\`

═══════════════════════════════════════════════════════════════
                    VALIDATION TASK
═══════════════════════════════════════════════════════════════

Analyze this program and report:
1. Whether the syntax is valid
2. Any structural errors (missing definitions, invalid references)
3. Warnings about potential issues or antipatterns
4. Suggestions for improvement

Respond with ONLY these XML tags:
<valid>true or false</valid>
<errors>
- error 1
- error 2
(or empty if no errors)
</errors>
<warnings>
- warning 1
- warning 2
(or empty if no warnings)
</warnings>
<summary>Brief summary of validation results</summary>`;

/**
 * Parse simple XML key-value pairs from text
 */
function parseKeyValueXml(text: string, keys: string[]): Record<string, string | undefined> {
  const result: Record<string, string | undefined> = {};
  for (const key of keys) {
    const regex = new RegExp(`<${key}>([\\s\\S]*?)</${key}>`, "i");
    const match = text.match(regex);
    if (match?.[1]) {
      const value = match[1].trim();
      if (value) {
        result[key] = value;
      }
    }
  }
  return result;
}

/**
 * Parse a list from text (lines starting with -)
 */
function parseList(text: string | undefined): string[] {
  if (!text) return [];
  return text
    .split("\n")
    .map((line) => line.trim())
    .filter((line) => line.startsWith("-"))
    .map((line) => line.substring(1).trim())
    .filter((line) => line.length > 0);
}

export const proseCompileAction: Action = {
  name: spec.name,
  description: spec.description,
  similes: spec.similes || [],
  examples: spec.examples
    ? spec.examples.map((ex) =>
        ex.map((msg) => ({
          name: msg.role,
          content: { text: msg.content },
        }))
      )
    : [],

  validate: async (runtime: IAgentRuntime, message: Memory, _state?: State): Promise<boolean> => {
    const content =
      typeof message.content === "string" ? message.content : message.content?.text || "";
    const lower = content.toLowerCase();

    // Match compile/validate commands
    if (lower.includes("prose compile")) return true;
    if (lower.includes("prose validate")) return true;
    if (lower.includes("check") && lower.includes(".prose")) return true;
    if (lower.includes("validate") && lower.includes(".prose")) return true;

    return false;
  },

  handler: async (
    runtime: IAgentRuntime,
    message: Memory,
    _state: State | undefined,
    _options: Record<string, unknown>,
    callback?: HandlerCallback
  ): Promise<boolean> => {
    const service = getService(runtime);
    const content =
      typeof message.content === "string" ? message.content : message.content?.text || "";

    try {
      // Extract file path
      const extraction = await generateText({
        runtime,
        context: EXTRACT_TEMPLATE.replace("{{content}}", content),
        modelType: ModelType.TEXT_SMALL,
      });

      const params = parseKeyValueXml(extraction, ["file"]);

      if (!params.file) {
        if (callback) {
          callback({
            text: "Please specify a .prose file to validate. Example: `prose compile workflow.prose`",
            actions: [],
          });
        }
        return false;
      }

      const file = params.file;
      const filePath = path.isAbsolute(file) ? file : path.join(process.cwd(), file);

      // Check if file exists
      const exists = await service.fileExists(filePath);
      if (!exists) {
        if (callback) {
          callback({
            text: `File not found: ${filePath}`,
            actions: [],
          });
        }
        return false;
      }

      // Read the program
      const programContent = await service.readProseFile(filePath);

      // Get compiler spec
      const compilerSpec =
        service.getCompilerSpec() ||
        "Validate the OpenProse program for syntax and structural correctness.";

      // Run validation
      const compilePrompt = COMPILE_TEMPLATE.replace("{{compiler_spec}}", compilerSpec).replace(
        "{{program}}",
        programContent
      );

      const validation = await generateText({
        runtime,
        context: compilePrompt,
        modelType: ModelType.TEXT_SMALL,
      });

      const result = parseKeyValueXml(validation, ["valid", "errors", "warnings", "summary"]);

      const isValid = result.valid?.toLowerCase() === "true";
      const errors = parseList(result.errors);
      const warnings = parseList(result.warnings);
      const summary = result.summary || "Validation complete.";

      logger.info(`[PROSE_COMPILE] Validated ${file}: valid=${isValid}`);

      // Build response
      const parts: string[] = [];
      parts.push(`## Validation Results for ${file}\n`);
      parts.push(`**Status:** ${isValid ? "✓ Valid" : "✗ Invalid"}\n`);
      parts.push(`**Summary:** ${summary}\n`);

      if (errors.length > 0) {
        parts.push("\n### Errors\n");
        for (const error of errors) {
          parts.push(`- ❌ ${error}`);
        }
      }

      if (warnings.length > 0) {
        parts.push("\n### Warnings\n");
        for (const warning of warnings) {
          parts.push(`- ⚠️ ${warning}`);
        }
      }

      if (isValid && errors.length === 0 && warnings.length === 0) {
        parts.push("\nNo issues found. Program is ready to run.");
      }

      if (callback) {
        callback({
          text: parts.join("\n"),
          actions: isValid ? ["PROSE_RUN"] : ["PROSE_COMPILE"],
          data: {
            valid: isValid,
            errors,
            warnings,
            file,
          },
        });
      }

      return true;
    } catch (error) {
      const errorMsg = error instanceof Error ? error.message : String(error);
      logger.error(`[PROSE_COMPILE] Error: ${errorMsg}`);

      if (callback) {
        callback({
          text: `Failed to validate program: ${errorMsg}`,
          actions: [],
        });
      }
      return false;
    }
  },
};
